File: frontends/app/src/features/workers/shareNotifications.test.ts
// Tests for the minimal share-notification adapter (SH-2 UX §5/§6): notice
// subscription/emission, badge counters derived from the canonical state
// (never cached as the source of truth), and the initial-sync rule (first
// load seeds the badge without toasting pre-existing access).
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import {
  onShareNotice, emitShareNotice, noticeFromEntry, shareNoticeMessage,
  syncSharedWithMe, markSharedSeen, resetShareNotifications,
  sharedWithMeCount, sharedUnreadCount,
} from './shareNotifications';
import type { SharedWithMeWorker } from './sharing';

function swm(id: string, sharedBy = 'ivan'): SharedWithMeWorker {
  return {
    worker_id: id, name: `Worker ${id}`, worker_type: 'audio', capabilities: null,
    owner_workspace_id: 'ws1', revoked_at: null, last_seen: null,
    created_at: null, granted_at: null,
    share_policy: { policy_id: 'p', scope_kind: 'users', starts_at: null, expires_at: null },
    access_reason: { kind: 'shared_by_user', shared_by: sharedBy, shared_by_display_name: null, owner_workspace_name: null },
  };
}

beforeEach(() => {
  resetShareNotifications();
});

// Signals are module-level singletons; the reset helper clears all session
// state (notice-dedup set, viewed set, both counters) between cases.
afterEach(() => {
  resetShareNotifications();
  vi.restoreAllMocks();
});

describe('notice transport seam', () => {
  it('subscribers receive emitted notices; unsubscribe stops them', () => {
    const seen: string[] = [];
    const off = onShareNotice((n) => seen.push(n.worker_name ?? ''));
    emitShareNotice({ event: 'worker.shared_with_user', worker_id: 'w1', worker_name: 'Home GPU', actor_username: 'ivan', ts: 1 });
    off();
    emitShareNotice({ event: 'worker.shared_with_user', worker_id: 'w1', worker_name: 'Again', actor_username: 'ivan', ts: 2 });
    expect(seen).toEqual(['Home GPU']);
  });

  it('a throwing subscriber never breaks others', () => {
    const ok = vi.fn();
    const off1 = onShareNotice(() => { throw new Error('boom'); });
    const off2 = onShareNotice(ok);
    emitShareNotice({ event: 'worker.shared_with_user', worker_id: 'w', worker_name: 'n', actor_username: 'a', ts: 1 });
    expect(ok).toHaveBeenCalledTimes(1);
    off1(); off2();
  });

  it('noticeFromEntry mirrors the backend event contract fields', () => {
    const n = noticeFromEntry(swm('w9', 'maria'), 42);
    expect(n).toEqual({
      event: 'worker.shared_with_user',
      worker_id: 'w9',
      worker_name: 'Worker w9',
      actor_username: 'maria',
      ts: 42,
    });
  });

  it('shareNoticeMessage formats «<username> shared Worker "X" with you»', () => {
    expect(shareNoticeMessage(
      { event: 'worker.shared_with_user', worker_id: 'w', worker_name: 'Home GPU', actor_username: 'Ivan', ts: 1 },
      (actor, worker) => `${actor} shared Worker "${worker}" with you`,
    )).toBe('Ivan shared Worker "Home GPU" with you');
  });
});

describe('syncSharedWithMe (badge + derived notices)', () => {
  it('initial sync seeds the count but raises NO notices', () => {
    const spy = vi.fn();
    const off = onShareNotice(spy);
    const fresh = syncSharedWithMe([], [swm('w1'), swm('w2')]);
    expect(fresh).toEqual([]);                 // no toasts for pre-existing access
    expect(sharedWithMeCount.value).toBe(2);   // badge reflects personal grants
    expect(sharedUnreadCount.value).toBe(2);
    expect(spy).not.toHaveBeenCalled();
    off();
  });

  it('a later sync with a new grant raises exactly one notice', () => {
    const spy = vi.fn();
    const off = onShareNotice(spy);
    syncSharedWithMe([], [swm('w1')]);
    const fresh = syncSharedWithMe([swm('w1')], [swm('w1'), swm('w2', 'maria')]);
    expect(fresh.map((w) => w.worker_id)).toEqual(['w2']);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0].actor_username).toBe('maria');
    off();
  });

  it('revoked/expired entries vanish from the count (server truth, not cache)', () => {
    syncSharedWithMe([], [swm('w1'), swm('w2')]);
    syncSharedWithMe([swm('w1'), swm('w2')], [swm('w1')]);
    expect(sharedWithMeCount.value).toBe(1);
  });

  it('badge counts personal shares only — the community pool never leaks in', () => {
    // The caller is responsible for feeding ONLY /workers/shared-with-me
    // results; this test pins the contract that the count equals the list.
    syncSharedWithMe([], [swm('w1')]);
    expect(sharedWithMeCount.value).toBe(1);
    syncSharedWithMe([swm('w1')], []);
    expect(sharedWithMeCount.value).toBe(0);
    expect(sharedUnreadCount.value).toBe(0);
  });

  it('markSharedSeen clears the unread badge but not the total', () => {
    const list = [swm('w1'), swm('w2')];
    syncSharedWithMe([], list);
    markSharedSeen(list);
    expect(sharedUnreadCount.value).toBe(0);
    expect(sharedWithMeCount.value).toBe(2);
  });

  it('markSharedSeen on an empty list is a no-op', () => {
    syncSharedWithMe([], []);
    markSharedSeen([]);
    expect(sharedUnreadCount.value).toBe(0);
  });
});

File: frontends/android/app/src/test/java/com/example/animastor/ui/ShareNotificationsTest.kt
package com.example.animastor.ui

import com.example.animastor.repository.SharedWithMeAccessReason
import com.example.animastor.repository.SharedWithMeWorker
import com.example.animastor.repository.SharePolicy
import org.junit.Assert.assertEquals
import org.junit.After
import org.junit.Before
import org.junit.Test

// Share notifications — minimal adapter (SH-2 UX §5/§6, web parity:
// frontends/app features/workers/shareNotifications.test.ts). Backend
// state is the source of truth; the module derives a session-only unread
// badge and a notice-transport seam subscribers can plug into.
class ShareNotificationsTest {

    private fun swm(
        id: String,
        sharedBy: String = "ivan"
    ): SharedWithMeWorker = SharedWithMeWorker(
        worker_id = id, name = "Worker $id", worker_type = "audio",
        capabilities = null, owner_workspace_id = "ws1", revoked_at = null,
        last_seen = null, created_at = null, granted_at = null,
        share_policy = SharePolicy(
            policy_id = "p", scope_kind = "users",
            starts_at = null, expires_at = null
        ),
        access_reason = SharedWithMeAccessReason(
            kind = "shared_by_user", shared_by = sharedBy,
            shared_by_display_name = null, owner_workspace_name = null
        )
    )

    @Before fun setUp() = ShareNotifications.resetShareNotifications()
    @After fun tearDown() = ShareNotifications.resetShareNotifications()

    // ── transport seam ──────────────────────────────────────────────────

    @Test fun subscribersReceiveEmittedNotices_unsubscribeStops() {
        val seen = mutableListOf<String?>()
        val off = ShareNotifications.onShareNotice { seen.add(it.worker_name) }
        ShareNotifications.emitShareNotice(ShareNotifications.ShareNotice(
            event = "worker.shared_with_user", worker_id = "w1",
            worker_name = "Home GPU", actor_username = "ivan", ts = 1,
        ))
        off()
        ShareNotifications.emitShareNotice(ShareNotifications.ShareNotice(
            event = "worker.shared_with_user", worker_id = "w1",
            worker_name = "Again", actor_username = "ivan", ts = 2,
        ))
        assertEquals(listOf("Home GPU"), seen)
    }

    @Test fun throwingSubscriberNeverBreaksOthers() {
        var okCalls = 0
        val off1 = ShareNotifications.onShareNotice { throw RuntimeException("boom") }
        val off2 = ShareNotifications.onShareNotice { okCalls++ }
        ShareNotifications.emitShareNotice(ShareNotifications.ShareNotice(
            "worker.shared_with_user", "w", "n", "a", 1,
        ))
        assertEquals(1, okCalls)
        off1(); off2()
    }

    @Test fun noticeFromEntryMirrorsEventContract() {
        val n = ShareNotifications.noticeFromEntry(swm("w9", "maria"), ts = 42)
        assertEquals(ShareNotifications.ShareNotice.EVENT_WORKER_SHARED, n.event)
        assertEquals("w9", n.worker_id)
        assertEquals("Worker w9", n.worker_name)
        assertEquals("maria", n.actor_username)
        assertEquals(42, n.ts)
    }

    // ── syncSharedWithMe (badge + derived notices) ─────────────────────

    @Test fun initialSyncSeedsCountWithoutToasting() {
        val spy = mutableListOf<ShareNotifications.ShareNotice>()
        val off = ShareNotifications.onShareNotice { spy.add(it) }
        val fresh = ShareNotifications.syncSharedWithMe(
            prev = emptyList(),
            next = listOf(swm("w1"), swm("w2")),
        )
        assertEquals(emptyList<SharedWithMeWorker>(), fresh)
        assertEquals(2, ShareNotifications.sharedWithMeCount)
        assertEquals(2, ShareNotifications.sharedUnreadCount)
        assertEquals(0, spy.size)
        off()
    }

    @Test fun laterSyncWithNewGrantRaisesExactlyOneNotice() {
        val spy = mutableListOf<ShareNotifications.ShareNotice>()
        val off = ShareNotifications.onShareNotice { spy.add(it) }
        ShareNotifications.syncSharedWithMe(emptyList(), listOf(swm("w1")))
        val fresh = ShareNotifications.syncSharedWithMe(
            prev = listOf(swm("w1")),
            next = listOf(swm("w1"), swm("w2", sharedBy = "maria")),
        )
        assertEquals(listOf("w2"), fresh.map { it.worker_id })
        assertEquals(1, spy.size)
        assertEquals("maria", spy[0].actor_username)
        // Badge: 2 entries total, 1 unseen (w2 is the fresh one).
        assertEquals(2, ShareNotifications.sharedWithMeCount)
        assertEquals(2, ShareNotifications.sharedUnreadCount)
        off()
    }

    @Test fun revokedOrExpiredEntriesVanishFromCount() {
        ShareNotifications.syncSharedWithMe(emptyList(), listOf(swm("w1"), swm("w2")))
        ShareNotifications.syncSharedWithMe(
            prev = listOf(swm("w1"), swm("w2")),
            next = listOf(swm("w1")),
        )
        assertEquals(1, ShareNotifications.sharedWithMeCount)
    }

    @Test fun markSharedSeenClearsUnreadBadgeButNotTotal() {
        val list = listOf(swm("w1"), swm("w2"))
        ShareNotifications.syncSharedWithMe(emptyList(), list)
        ShareNotifications.markSharedSeen(list)
        assertEquals(0, ShareNotifications.sharedUnreadCount)
        assertEquals(2, ShareNotifications.sharedWithMeCount)
    }

    @Test fun markSharedSeenOnEmptyListIsNoop() {
        ShareNotifications.syncSharedWithMe(emptyList(), emptyList())
        ShareNotifications.markSharedSeen(emptyList())
        assertEquals(0, ShareNotifications.sharedUnreadCount)
    }
}
